basyx/aas/model/base.py

```diff
@@ -1029,48 +1029,30 @@ def resolve(self, provider_: "provider.AbstractObjectProvider") -> _RT:
                                      object is stored in the ``value`` attribute of the exception
         :raises KeyError: If the reference could not be resolved
         """
 
-        from . import SubmodelElementList
-
         # For ModelReferences, the first key must be an AasIdentifiable. So resolve the first key via the provider.
         identifier: Optional[Identifier] = self.key[0].get_identifier()
         if identifier is None:
             raise AssertionError("Retrieving the identifier of the first key failed.")
 
-        resolved_keys: List[str] = []  # for more helpful error messages
         try:
             item: Referable = provider_.get_identifiable(identifier)
         except KeyError as e:
             raise KeyError("Could not resolve identifier {}".format(identifier)) from e
-        resolved_keys.append(str(identifier))
 
-        # All keys following the first must not reference identifiables (AASd-125). Thus, we can just follow the path
-        # recursively.
-        for key in self.key[1:]:
-            if not isinstance(item, UniqueIdShortNamespace):
-                raise TypeError("Object retrieved at {} is not a Namespace".format(" / ".join(resolved_keys)))
-            is_submodel_element_list = isinstance(item, SubmodelElementList)
-            try:
-                if is_submodel_element_list:
-                    # The key's value must be numeric, since this is checked for keys following keys of type
-                    # SUBMODEL_ELEMENT_LIST on construction of ModelReferences.
-                    # Additionally item is known to be a SubmodelElementList which supports __getitem__ because we're in
-                    # the `is_submodel_element_list` branch, but mypy doesn't infer types based on isinstance checks
-                    # stored in boolean variables.
-                    item = item.value[int(key.value)]  # type: ignore
-                    resolved_keys[-1] += f"[{key.value}]"
-                else:
-                    item = item.get_referable(key.value)
-                    resolved_keys.append(item.id_short)
-            except (KeyError, IndexError) as e:
-                raise KeyError("Could not resolve {} {} at {}".format(
-                    "index" if is_submodel_element_list else "id_short", key.value, " / ".join(resolved_keys)))\
-                    from e
+        # All keys following the first must not reference identifiables (AASd-125). Thus, we can just resolve the
+        # id_short path via get_referable().
+        # This is cursed af, but at least it keeps the code DRY. get_referable() will check the type of self in the
+        # first iteration, so we can ignore the type here.
+        item = UniqueIdShortNamespace.get_referable(item,  # type: ignore[arg-type]
+                                                    map(lambda k: k.value, self.key[1:]))
```

Comment:
This is very elegant but indeed really hard to understand.

Reply:
I think the `map(lambda...` expression is pretty clean, but calling `get_referable()` and manually supplying `self` as `item`, because `item` is not necessarily a `UniqueIdShortNamespace`, is not. However, I think it's acceptable because `get_referable()` checks the type of `self` in the first iteration and because it's all documented via comments, in `ModelReference.resolve()` as well as `UniqueIdShortNamespace.get_referable()`. It's the only solution I could come up with that doesn't require duplicate error messages, i.e. having the same check with the same error message in both functions.